client/components/jetpack/licensing-prompt-dialog/index.tsx

```diff
@@ -0,0 +1,113 @@
+import { Button, Dialog, Gridicon } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import { preventWidows } from 'calypso/lib/formatting';
+import { recordTracksEvent } from 'calypso/state/analytics/actions';
+import { getSiteAdminUrl } from 'calypso/state/sites/selectors';
+import { getUserLicenses, getUserLicensesCounts } from 'calypso/state/user-licensing/selectors';
+import type { License } from 'calypso/state/user-licensing/types';
+
+import './style.scss';
+
+interface Props {
+	siteId: number;
+}
+
+function LicensingPromptDialog( { siteId }: Props ) {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	const [ showLicensesDialog, setShowLicensesDialog ] = useState< boolean >( false );
+	const [ detachedUserLicense, setDetachedUserLicense ] = useState< License | null >( null );
+	const userLicenses = useSelector( getUserLicenses );
+	const userLicensesCounts = useSelector( getUserLicensesCounts );
+	const siteAdminUrl = useSelector( ( state ) => getSiteAdminUrl( state, siteId ) );
+	const jetpackDashboardUrl = siteAdminUrl + 'admin.php?page=jetpack#/my-plan';
+
+	const hasOneDetachedLicense = userLicensesCounts && userLicensesCounts[ 'detached' ] === 1;
+	const hasDetachedLicenses = userLicensesCounts && userLicensesCounts[ 'detached' ] !== 0;
+
+	useEffect( () => {
+		if ( userLicenses && hasOneDetachedLicense ) {
+			setDetachedUserLicense(
+				Object.values( userLicenses.items ).filter( ( { attachedAt } ) => attachedAt === null )[ 0 ]
+			);
+		}
+	}, [ hasOneDetachedLicense, userLicenses ] );
+
+	useEffect( () => {
+		if ( hasDetachedLicenses && siteAdminUrl ) {
+			setShowLicensesDialog( true );
+		}
+	}, [ hasDetachedLicenses, siteAdminUrl ] );
+
+	useEffect( () => {
+		dispatch( recordTracksEvent( 'calypso_user_license_modal_view' ) );
+	}, [ dispatch ] );
+
+	const title = useMemo( () => {
+		if ( hasOneDetachedLicense ) {
+			return preventWidows(
+				translate( 'Your %(productName)s is pending activation', {
+					args: {
+						productName: detachedUserLicense && detachedUserLicense.product,
+					},
+				} )
+			);
+		}
+		return preventWidows( translate( 'You have an available product license key' ) );
+	}, [ detachedUserLicense, hasOneDetachedLicense, translate ] );
+
+	const activateProductClick = useCallback( () => {
+		dispatch( recordTracksEvent( 'calypso_user_license_modal_activate_click' ) );
+	}, [ dispatch ] );
+
+	const selectAnotherProductClick = useCallback( () => {
+		setShowLicensesDialog( false );
+		dispatch( recordTracksEvent( 'calypso_user_license_modal_close_click' ) );
+	}, [ dispatch ] );
+
+	return (
+		<Dialog
+			additionalClassNames="licensing-prompt-dialog"
+			isVisible={ showLicensesDialog }
+			onClose={ () => setShowLicensesDialog( false ) }
```

Comment:
What about wrapping this callback inside `useCallback` to avoid re-renders?

Reply:
Good call. I have updated the PR and wrapped that function call with `useCallback`.

Reply:
Thanks